engine: fix intent detection in MVCCIncrementalIterator

dt opened the pull request:
looks like we flipped the operands compared to when mvccGetIternal would return an intent err.
we missed it in the mvcc test since we accidentally closed over the wrong time bounds.

Fixes #13189
Fixes #13016
Fixes #13600

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13771)
<!-- Reviewable:end -->

dt committed:
sqlccl: remove `SucceedsSoon` from backup "Bank" test

For waiting for bg workers to make progress, replaces rehashing of the whole
table in a `SucceedsSoon` with a signal channel. This simplification seems to
speed the test up a bit in stress.

dt committed:
engine: fix intent detection in mvccincrementaliter

looks like we flipped the operands compared to when mvccGetIternal would return an intent err.
we missed it in the mvcc test since we accidentally closed over the wrong time bounds.

Fixes #13189
Fixes #13016
Fixes #13600

tamird commented:
Reviewed 1 of 1 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup_test.go, line 265 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgfq6XtfyDgqRyBc3D:-Kdgfq6YnAt8pQteSk6G:b-hndldg) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c10232ce73c49cf1b73e72dba77af334eac721/pkg/ccl/sqlccl/backup_test.go#L265)):*
> ```Go
> 
> func startBackgroundWrites(
> 	stopper *stop.Stopper, sqlDB *gosql.DB, wake chan struct{}, maxID int,
> ```

chan<- (or however it's spelled)

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgg9ABJqzpeyM3wWrf:-Kdgg9ABJqzpeyM3wWrg:b-v98dir) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c10232ce73c49cf1b73e72dba77af334eac721/pkg/ccl/sqlccl/backup_test.go#L340)):*
> ```Go
> 
> 		// Set the id=balance invariant on each row and back up the table.
> 		_ = sqlDB.Exec(`UPDATE bench.bank SET balance = id`)
> ```

does this return an error? why not check it? if it doesn't, why the suppression-style assignment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/sqlccl/backup_test.go, line 265 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgfq6XtfyDgqRyBc3D:-KdghanpwvOIXLmturfi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c10232ce73c49cf1b73e72dba77af334eac721/pkg/ccl/sqlccl/backup_test.go#L265)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

chan<- (or however it's spelled)
</blockquote></details>

Done.

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgg9ABJqzpeyM3wWrf:-KdggerGITd_wAERIP15:b-9dyf0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c10232ce73c49cf1b73e72dba77af334eac721/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this return an error? why not check it? if it doesn't, why the suppression-style assignment?
</blockquote></details>

returns a `gosql.Result`. Exec is a wrapper around gosql.Exec that kills the test on error.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/ccl/sqlccl/backup_test.go, line 340 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgg9ABJqzpeyM3wWrf:-Kdghj4VnPKrKYR0zCst:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6c10232ce73c49cf1b73e72dba77af334eac721/pkg/ccl/sqlccl/backup_test.go#L340)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

returns a `gosql.Result`. Exec is a wrapper around gosql.Exec that kills the test on error.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

danhhz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

the first commit's message seems out of date, though

---

Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-:-KdgiS_YoSpV1hKh_j1d:bhewngn)*
<!-- Sent from Reviewable.io -->

dt commented:
Fixed.

tamird commented:
Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Looks like this might also have fixed #13735:
```
make stress PKG=./pkg/ccl/storageccl TESTS=Export
...
40213 runs so far, 0 failures, over 20m30s
```

danhhz commented:
hallelujah. throw a "Closes #13735" in the commit message?

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/mvcc.go, line 19 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgnbt-TU4CFYdgKuvj:-Kdgnbt-TU4CFYdgKuvk:bvjt8vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/47dcc75b370c2f7edec75a86272a9994acf28354/pkg/ccl/storageccl/engineccl/mvcc.go#L19)):*
> ```Go
> )
> 
> // MVCCIncrementalIterator iterates over the diff of the key range [start,end)
> ```

mind using the right members for the key range and time range? Otherwise this comment is very confusing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-:-KdgoqvGtOJZXSTowguz:bnfp4nl)*
<!-- Sent from Reviewable.io -->

dt commented:
@danhhz hmm, on `master` i also get `45083 runs so far, 0 failures, over 22m45s` so I'm withdrawing the `Fixes #13735` for now

danhhz commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 88 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdiCwBpd8kuJh7yRg5m:-KdiCwBqf7NrjbMk2siW:bg4opyd) ([raw file](https://github.com/cockroachdb/cockroach/blob/65a383f30236f5c457a1547d9d893c3bce618bff/pkg/ccl/storageccl/engineccl/mvcc_test.go#L88)):*
> ```Go
> 			iter := NewMVCCIncrementalIterator(e)
> 			defer iter.Close()
> 			for iter.Reset(keyMin, keyMax, startTime, endTime); iter.Valid(); iter.Next() {
> ```

this should also be startKey and endKey instead of keyMin and keyMax

I'm pretty sure the entire var block above can be moved below these funcs which should catch any others lurking about

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/mvcc.go, line 19 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdgnbt-TU4CFYdgKuvj:-KdkpNala97KNKjBm7Li:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/21923a94dc228ae064a3d17549da12bbd3f9ee21/pkg/ccl/storageccl/engineccl/mvcc.go#L19)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

mind using the right members for the key range and time range? Otherwise this comment is very confusing.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 88 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdiCwBpd8kuJh7yRg5m:-KdkpOHNSzUB7kp_LXqx:b-1vd3zi) ([raw file](https://github.com/cockroachdb/cockroach/blob/65a383f30236f5c457a1547d9d893c3bce618bff/pkg/ccl/storageccl/engineccl/mvcc_test.go#L88)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this should also be startKey and endKey instead of keyMin and keyMax

I'm pretty sure the entire var block above can be moved below these funcs which should catch any others lurking about
</blockquote></details>

bah, yep, and starts failing too when fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt committed:
engineccl: pull subtest helpers out, name all subtests

pulling the subtest helpers out makes closing over or shadowing vars impossible and makes it easier (for me) to reason about what they're actually using and testing.

prior to this the iterate test was always using min and max rather than the passed key bounds.
after switching bounds, note that the two iterate test calls need to PrefixEnd their end arg.

dt commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 88 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdiCwBpd8kuJh7yRg5m:-KdkunWnVOYVcWmdF4p6:b3j5w6x) ([raw file](https://github.com/cockroachdb/cockroach/blob/65a383f30236f5c457a1547d9d893c3bce618bff/pkg/ccl/storageccl/engineccl/mvcc_test.go#L88)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

bah, yep, and starts failing too when fixed.
</blockquote></details>

Okay, pulled into first-class funcs, which can't close over or shadow anything, making it clearer, imo, what they actually use and test.

Went ahead and named all the subtests too so we can tell which one fails.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 163 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkvVuCVrJQRcRioCFy:-KdkvVuCVrJQRcRioCFz:bpwpgrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L163)):*
> ```Go
> 	}
> 	t.Run("intents1",
> 		iterateExpectErr(e, testKey1, testKey1.PrefixEnd(), ts0, tsMax, "conflicting intents"))
> ```

NB: after fixing the start and end bounds in subtest refactor, we need to `PrefixEnd` these end keys to pass, but I think that is correct? cc @danhhz

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 3 files at r6, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 116 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkwzdzP6mAhDfgvBZB:-KdkwzdzP6mAhDfgvBZC:b9oo1z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L116)):*
> ```Go
> 
> 	// Exercise time ranges.
> 	t.Run("ts 1-1", assertEqualKVs(e, keyMin, keyMax, ts1, ts1, nil))
> ```

are these both repeated below?

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 120 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdkx4_6Ew_75PmIv5kt:-Kdkx4_6Ew_75PmIv5ku:bqs6y7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L120)):*
> ```Go
> 
> 	// Exercise time ranges.
> 	t.Run("ts 1-1", assertEqualKVs(e, keyMin, keyMax, ts1, ts1, nil))
> ```

the first two of these are repeated?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 120 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-Kdkx4_6Ew_75PmIv5kt:-KdkxOUaITNkBaTVmmp9:b6p28hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

the first two of these are repeated?
</blockquote></details>

indeed, fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

danhhz commented:
Reviewed 2 of 3 files at r6, 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 101 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkxRyobdkLfrsdCMDs:-KdkxRypWWGYv2bEmUfU:b-65n8f0) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L101)):*
> ```Go
> 	kv1_1_1 := makeKVT(testKey1, testValue1, ts1)
> 	kv1_4_4 := makeKVT(testKey1, testValue4, ts4)
> 	kv2_2_1 := makeKVT(testKey1, testValue2, ts2)
> ```

this one is named wrong. should be kv1_2_2 (confused me when reading "kv 1-2")

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 163 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkvVuCVrJQRcRioCFy:-Kdkxu2_yYf9eJdam55K:b-yfky04) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L163)):*
<details><summary><i>Previously, dt (David Taylor) wrote…</i></summary><blockquote>

NB: after fixing the start and end bounds in subtest refactor, we need to `PrefixEnd` these end keys to pass, but I think that is correct? cc @danhhz
</blockquote></details>

yeah, i think PrefixEnd is correct

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->

dt commented:
Review status: 2 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 101 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkxRyobdkLfrsdCMDs:-Kdl3DBpKl9FPzl58LHi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L101)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

this one is named wrong. should be kv1_2_2 (confused me when reading "kv 1-2")
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/mvcc_test.go, line 116 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771#-KdkwzdzP6mAhDfgvBZB:-KdkyN37fkOJtp-gFAAC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/02ff46fb4cf613073d5010cfba99eb1dd15a16d4/pkg/ccl/storageccl/engineccl/mvcc_test.go#L116)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are these both repeated below?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13771)*
<!-- Sent from Reviewable.io -->
